service-base/src/main/java/org/eclipse/hono/service/plan/PrometheusBasedResourceLimitChecks.java

```diff
@@ -125,6 +131,52 @@ public PrometheusBasedResourceLimitChecks(final WebClient webClient) {
         }
     }
 
+    /**
+     * Checks if accepting the incoming message exceeds the configured data volume limit for a tenant or not.
+     *
+     * @param tenant The tenant configuration to check the limit against.
+     * @param payloadSize The message payload size in bytes.
+     * @return A future indicating the outcome of the check.
+     *         <p>
+     *         The future will be failed with a {@link ServiceInvocationException} if the check could not be performed.
+     */
+    public Future<Boolean> isMessageLimitExceeded(final TenantObject tenant,
+            final long payloadSize) {
+
+        Objects.requireNonNull(tenant);
+        final long maxBytes = getMaximumNumberOfBytes(tenant);
+        final LocalDate effectiveSince = getEffectiveSince(tenant);
+        final long periodInDays = getPeriodInDays(tenant);
+
+        log.trace("message limit config for tenant [{}] are [{}:{}, {}:{}, {}:{}]", tenant.getTenantId(),
+                FIELD_MAX_BYTES, maxBytes, FIELD_EFFECTIVE_SINCE, effectiveSince, FIELD_PERIOD_IN_DAYS, periodInDays);
+
+        if (maxBytes == -1 || effectiveSince == null || periodInDays <= 0 || payloadSize <= 0) {
+            return Future.succeededFuture(Boolean.FALSE);
+        } else {
+            final long dataUsagePeriod = calculateDataUsagePeriod(effectiveSince, periodInDays);
+            if (dataUsagePeriod <= 0) {
+                return Future.succeededFuture(Boolean.FALSE);
+            }
+            final String queryParams = String.format("floor(sum(increase(%s{status=\"%s\", tenant=\"%s\"} [%sd])))",
+                    MESSAGES_PAYLOAD_SIZE_METRIC_NAME,
+                    MetricsTags.ProcessingOutcome.FORWARDED.asTag().getValue(),
+                    tenant.getTenantId(),
+                    dataUsagePeriod);
+            return executeQuery(queryParams)
```

Comment:
we cannot execute this query for every message published by a device, because this is waaayyy too expensive to do. Instead, we could maintain a local counter which gets updated with the current message's payload size and which is periodically updated based on the value retrieved from the metrics back end.

Reply:
It is already intended for the next PR while equipping the _protocol adapters_ with this new method.
 I thought of a caching mechanism similar to the tenant and registration clients. Then the user can configure the cache timeout according to his/her needs. The method `isMessageLimitReached` is not aware, if the message will be later successfully accepted or not.  Where as protocol adapters know and the delta part can be updated by the protocol adapters to the cache. The query result can be updated by the `PrometheusBasedResourceLimitChecks`. WDYT?

Reply:
IMHO a simple timer running the query towards the metrics backend periodically would do ...

Reply:
Given a scenario: A device `D1` belonging to a tenant `Tenant1` sends an _event_ message. The metrics back-end is queried for the _consumed bytes_ for the tenant `Tenant1`. And a timer is set periodically to retrieve this data for `Tenant1`. Assume the next message from this device `D1` is 3 days later or even later. This timer periodically queries the metrics back-end, which IMHO is not needed. The number of timers will then be equal to the number of _tenants_ of the connected devices that have sent at-least one message.

IMHO a data structure to store the _consumed bytes_ grouped by tenant id is needed and should be configurable to say how long the old data can be used. Hence I think that caching mechanism may suit the purpose.

Reply:
I agree, that would be even better. All I am saying is that the code in this PR should not retrieve the value for each request ...

Reply:
I agree and it is intended as part of the next PR.

Reply:
I understood that this is your plan. However, I will not approve this PR with the query being executed on every message. So you can either implement the caching in this PR already or (much easier) refactor the code executing the query into a separate method which is never invoked (and then always return `Boolean.FALSE`).

Reply:
IMHO currently no protocol adapters are equipped with the method `isMessageLimitReached(...)` to check the message limit. On the contrary, this query is currently never being executed. If you think its still better to have caching mechanism or separate method for the query which is never invoked as part of this PR, I will add to this.

Reply:
please do